File: DataDriven/expressJS/src/controllers/Author.js
const Author = require("../models/Author");
const Post = require("../models/Post");

exports.getPosts = async (req, res) => {
  try {
    const { emails } = req.body;
    const authors = await Author.find({ email: { $in: emails } }).exec();
    if (authors.length < 1) {
      return res.status(200).json({ result: "No authors found." });
    }
    const authorIDs = authors.map(({ _id }) => _id);
    const postsByAuthor = await Post.find({
      authors: { $in: authorIDs },
    }).exec();
    if (postsByAuthor.length < 1) {
      res.status(200).json({ result: "No posts found." });
    } else {
      res.status(200).json({ result: postsByAuthor });
    }
  } catch (e) {
    res.status(500).json({ error: `${e}` });
  }
};

File: DataDriven/Gin/internal/server/controllers/author.go
package controllers

import (
	"fmt"
	"net/http"

	"Gin/internal/models"

	"github.com/gin-gonic/gin"
)

func (s *Server) GetPostsByAuthorHandler(c *gin.Context) {

	var database = s.db
	var req struct {
		Emails []string `json:"emails"`
	}
	fmt.Println("ddd")

	if err := c.BindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var authors []models.Author
	if err := database.Where("email IN (?)", req.Emails).Find(&authors).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(authors) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No authors found."})
		return
	}

	var authorIDs []uint
	for _, author := range authors {
		authorIDs = append(authorIDs, author.ID)
	}

	var posts []models.Post
	if err := database.
		Joins("JOIN post_authors pa ON pa.post_id = posts.id").
		Where("pa.author_id IN (?)", authorIDs).Preload("Authors").
		Find(&posts).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if len(posts) < 1 {
		c.JSON(http.StatusOK, gin.H{"result": "No posts found."})
	} else {
		c.JSON(http.StatusOK, gin.H{"result": posts})
	}
}
